src/image.rs

```diff
@@ -0,0 +1,228 @@
+                "kCGColorSpaceGenericRGB" => {
+                    match self.alphainfo().unwrap() {
+                        ImageAlphaInfo::None
+                        | ImageAlphaInfo::NoneSkipLast
+                        | ImageAlphaInfo::NoneSkipFirst => {
+                            Ok(ColorSpace::RGB)
+                        },
+                        ImageAlphaInfo::PremultipliedLast
+                        | ImageAlphaInfo::Last => {
+                            Ok(ColorSpace::RGBA)
+                        },
+                        ImageAlphaInfo::PremultipliedFirst
+                        | ImageAlphaInfo::First => {
+                            Ok(ColorSpace::ARGB)
+                        },
+                        _ => {
+                            Err("Unsupport")
+                        }
+                    }
+                },
+                "kCGColorSpaceITUR_709" => {
+                    match self.alphainfo().unwrap() {
+                        ImageAlphaInfo::None
+                        | ImageAlphaInfo::NoneSkipLast
+                        | ImageAlphaInfo::NoneSkipFirst => {
+                            Ok(ColorSpace::YCbCr)
+                        },
+                        _ => {
+                            Err("Unsupport")
+                        }
+                    }
+                },
+                "kCGColorSpaceITUR_2020" => {
+                    match self.alphainfo().unwrap() {
+                        ImageAlphaInfo::None
+                        | ImageAlphaInfo::NoneSkipLast
+                        | ImageAlphaInfo::NoneSkipFirst => {
+                            Ok(ColorSpace::YCbCr)
+                        },
+                        _ => {
+                            Err("Unsupport")
+                        }
+                    }
+                },
+                // TODO: kCGColorSpaceGenericGray, kCGColorSpaceGenericCMYK ....
+                _ => {
+                    Err("Unsupport")
+                }
+            }
+            
+        }
+    }
+    pub fn alphainfo(&self) -> Result<ImageAlphaInfo, &'static str> {
+        unsafe {
+            ImageAlphaInfo::from_u32(CGImageGetAlphaInfo(self.0))
+        }
+    }
+    pub fn raw_len(&self) -> usize {
+        self.width() * self.height() * self.pixel_width()
+    }
+    pub fn data(&self) -> &[u8] {
+        // Image Pixels Data (BGRA)
+        unsafe {
+            let cf_data = CGDataProviderCopyData(CGImageGetDataProvider(self.0));
+            let raw_len = CFDataGetLength(cf_data) as usize;
+            // Image size is inconsistent with W*H*D.
+            assert_eq!(self.width()*self.height()*self.pixel_bits(), raw_len*8 );
+            assert_eq!(self.raw_len(), raw_len );
+
+            let data = slice::from_raw_parts(CFDataGetBytePtr(cf_data), raw_len);
+            CFRelease(cf_data as *const c_void);
+            // WARN: Slow.
+            // let v = data.to_vec();
+            data
```

Comment:
Are you sure that the implementation of this function is correct? It seems to me that you're returning a slice pointing to the invalid data. Remember that `slice` does not own the data you passed to it

Basically you created the slice which points to the data you copied from the data provider, and right after that (but before returning the slice itself) you release the data.